File: mobile/__tests__/workshopLive.test.ts
/**
 * Pure-logic tests for lib/workshopLive.ts (technician live-status engine).
 * Deterministic: every call passes an explicit numeric `now`.
 */
import {
  statusFromEvents,
  isCheckedIn,
  myProductivityToday,
  statusKind,
  statusLabel,
  WorkshopEventLike,
} from '../lib/workshopLive'

const MIN = 60_000
const ev = (event_type: string, atMin: number, reason_code?: string): WorkshopEventLike => ({
  event_type,
  reason_code: reason_code ?? null,
  at: atMin * MIN,
})

describe('statusFromEvents', () => {
  it('defaults to available when present and absent when not, with no events', () => {
    expect(statusFromEvents([], { present: true })).toBe('available')
    expect(statusFromEvents([], { present: false })).toBe('absent')
    expect(statusFromEvents(null, {})).toBe('absent')
  })

  it('reflects the latest duty/job event', () => {
    expect(statusFromEvents([ev('check_in', 0), ev('start_job', 10)])).toBe('working')
    expect(statusFromEvents([ev('start_job', 0), ev('start_break', 5)])).toBe('on_break')
    expect(statusFromEvents([ev('start_job', 0), ev('complete_task', 5)])).toBe('awaiting_inspection')
    expect(statusFromEvents([ev('check_in', 0), ev('check_out', 30)])).toBe('off_duty')
  })

  it('resolves a pause_job by its blocked reason_code', () => {
    expect(statusFromEvents([ev('start_job', 0), ev('pause_job', 5, 'parts')])).toBe('waiting_parts')
    expect(statusFromEvents([ev('start_job', 0), ev('pause_job', 5, 'tools')])).toBe('waiting_tools')
    expect(statusFromEvents([ev('start_job', 0), ev('pause_job', 5, 'break')])).toBe('on_break')
    // Unknown reason falls back to available.
    expect(statusFromEvents([ev('start_job', 0), ev('pause_job', 5, 'other')])).toBe('available')
  })

  it('treats request_assistance / report_problem as annotations that do not change status', () => {
    expect(
      statusFromEvents([ev('start_job', 0), ev('request_assistance', 5), ev('report_problem', 6)]),
    ).toBe('working')
  })
})

describe('isCheckedIn', () => {
  it('is true after check_in and false after check_out', () => {
    expect(isCheckedIn([ev('check_in', 0)])).toBe(true)
    expect(isCheckedIn([ev('check_in', 0), ev('check_out', 60)])).toBe(false)
    // Job events between check_in/out do not toggle duty.
    expect(isCheckedIn([ev('check_in', 0), ev('start_job', 10)])).toBe(true)
  })

  it('is false with no duty events', () => {
    expect(isCheckedIn([])).toBe(false)
    expect(isCheckedIn(null)).toBe(false)
  })
})

describe('myProductivityToday', () => {
  it('splits duty into productive / blocked / unassigned and counts completed jobs', () => {
    // start_job@0 -> productive until request_parts@60 (60 min productive)
    // request_parts@60 -> blocked:parts until resume_job@90 (30 min blocked)
    // resume_job@90 -> productive until complete_task@150 (60 min productive)
    const events = [
      ev('start_job', 0),
      ev('request_parts', 60),
      ev('resume_job', 90),
      ev('complete_task', 150),
    ]
    const p = myProductivityToday(events, { now: 150 * MIN })
    expect(p.productiveMin).toBe(120)
    expect(p.blockedMin).toBe(30)
    expect(p.breakMin).toBe(0)
    expect(p.unassignedMin).toBe(0)
    expect(p.jobsCompleted).toBe(1)
  })

  it('returns zeroed metrics for no events', () => {
    const p = myProductivityToday([], { now: 100 * MIN })
    expect(p).toEqual({
      productiveMin: 0,
      blockedMin: 0,
      unassignedMin: 0,
      breakMin: 0,
      jobsCompleted: 0,
    })
  })
})

describe('status presentation helpers', () => {
  it('maps status to a theme StatusKind', () => {
    expect(statusKind('working')).toBe('success')
    expect(statusKind('waiting_parts')).toBe('warning')
    expect(statusKind('absent')).toBe('danger')
    expect(statusKind('off_duty')).toBe('neutral')
  })

  it('gives a human label, defaulting safely', () => {
    expect(statusLabel('working')).toBe('Working')
    expect(statusLabel('waiting_approval')).toBe('Waiting for Approval')
  })
})

File: tyre_pulse_flutter/test/features/workshop/domain/workshop_live_test.dart
// Parity with mobile/lib/workshopLive.ts (statusFromEvents, isCheckedIn,
// myProductivityToday) and the V291 event vocabulary.
import 'package:flutter_test/flutter_test.dart';
import 'package:tyre_pulse/features/workshop/data/workshop_repository.dart';
import 'package:tyre_pulse/features/workshop/domain/workshop_live.dart';

WorkshopEventLike _e(String type, int minute, {String? reason}) =>
    WorkshopEventLike(
      eventType: type,
      reasonCode: reason,
      at: DateTime.utc(2026, 9, 26, 8).add(Duration(minutes: minute)),
    );

void main() {
  group('workshopStatusFromEvents', () {
    test('no events: available when present, else absent', () {
      expect(
        workshopStatusFromEvents(const <WorkshopEventLike>[], present: true),
        WorkshopStatus.available,
      );
      expect(
        workshopStatusFromEvents(const <WorkshopEventLike>[]),
        WorkshopStatus.absent,
      );
    });

    test('the latest meaningful event decides, by time not list order', () {
      expect(
        workshopStatusFromEvents(<WorkshopEventLike>[
          _e('request_parts', 30),
          _e('start_job', 10),
        ]),
        WorkshopStatus.waitingParts,
      );
      expect(
        workshopStatusFromEvents(<WorkshopEventLike>[_e('complete_task', 5)]),
        WorkshopStatus.awaitingInspection,
      );
      expect(
        workshopStatusFromEvents(<WorkshopEventLike>[_e('check_out', 5)]),
        WorkshopStatus.offDuty,
      );
    });

    test('annotations never change the status', () {
      expect(
        workshopStatusFromEvents(<WorkshopEventLike>[
          _e('start_job', 0),
          _e('report_problem', 5),
          _e('request_assistance', 6),
        ]),
        WorkshopStatus.working,
      );
    });

    test('pause_job resolves by its reason code', () {
      WorkshopStatus pause(String? r) => workshopStatusFromEvents(
            <WorkshopEventLike>[_e('pause_job', 0, reason: r)],
          );
      expect(pause('parts'), WorkshopStatus.waitingParts);
      expect(pause('tools'), WorkshopStatus.waitingTools);
      expect(pause('approval'), WorkshopStatus.waitingApproval);
      expect(pause('vehicle'), WorkshopStatus.waitingVehicle);
      expect(pause('break'), WorkshopStatus.onBreak);
      expect(pause(null), WorkshopStatus.available);
    });
  });

  test('workshopIsCheckedIn reads only duty events', () {
    expect(
      workshopIsCheckedIn(<WorkshopEventLike>[
        _e('check_in', 0),
        _e('start_job', 5),
      ]),
      isTrue,
    );
    expect(
      workshopIsCheckedIn(<WorkshopEventLike>[
        _e('check_in', 0),
        _e('check_out', 60),
      ]),
      isFalse,
    );
    expect(workshopIsCheckedIn(const <WorkshopEventLike>[]), isFalse);
  });

  test('productivity splits productive, blocked and break time', () {
    final WorkshopProductivity p = workshopProductivityToday(
      <WorkshopEventLike>[
        _e('check_in', 0),
        _e('start_job', 10),
        _e('request_parts', 70),
        _e('resume_job', 100),
        _e('start_break', 130),
        _e('end_break', 145),
        _e('complete_task', 160),
        _e('check_out', 170),
      ],
      now: DateTime.utc(2026, 9, 26, 12),
    );
    expect(p.productiveMin, 90);
    expect(p.blockedMin, 30);
    expect(p.breakMin, 15);
    expect(p.unassignedMin, 35);
    expect(p.jobsCompleted, 1);
  });

  test('every action is in the V291 CHECK vocabulary', () {
    for (final WorkshopTechAction a in kWorkshopTechActions) {
      expect(kWorkshopEventTypes, contains(a.event));
    }
    expect(
      kWorkshopTechActions
          .where((WorkshopTechAction a) => a.needsNote)
          .map((WorkshopTechAction a) => a.key),
      <String>[
        'request_parts',
        'request_assistance',
        'waiting_approval',
        'waiting_vehicle',
        'waiting_tools',
        'report_problem',
      ],
    );
  });

  test('formatWorkshopMinutes', () {
    expect(formatWorkshopMinutes(0), '0m');
    expect(formatWorkshopMinutes(45), '45m');
    expect(formatWorkshopMinutes(80), '1h 20m');
    expect(formatWorkshopMinutes(120), '2h');
    expect(formatWorkshopMinutes(-5), '0m');
  });

  test('isOpenWorkshopJob', () {
    expect(isOpenWorkshopJob(null), isTrue);
    expect(isOpenWorkshopJob('In Progress'), isTrue);
    expect(isOpenWorkshopJob('Completed'), isFalse);
    expect(isOpenWorkshopJob(' closed '), isFalse);
  });

  test('a synced row replaces its local copy by client_uuid', () {
    final List<WorkshopEventRecord> merged = mergeWorkshopEvents(
      <WorkshopEventRecord>[
        WorkshopEventRecord(
          eventType: 'start_job',
          clientUuid: 'ws_a',
          at: DateTime.utc(2026, 9, 26, 8),
        ),
      ],
      <WorkshopEventRecord>[
        WorkshopEventRecord(
          eventType: 'start_job',
          clientUuid: 'ws_a',
          localAt: DateTime.utc(2026, 9, 26, 8),
        ),
        WorkshopEventRecord(
          eventType: 'pause_job',
          clientUuid: 'ws_b',
          localAt: DateTime.utc(2026, 9, 26, 9),
        ),
      ],
    );
    expect(merged.map((WorkshopEventRecord e) => e.clientUuid), <String>[
      'ws_a',
      'ws_b',
    ]);
  });
}
